control-plane/csi-driver/src/filesystem.rs

```diff
@@ -0,0 +1,217 @@
+//! This module consists of the various filesystem specific operations utility. Including creation
+//! of filesystem, changing the parameter of filesystem like uuid, default mount options for
+//! specific filesystem, repairing of the filesystem, retrieving specific properties of the
+//! filesystem.
+
+use crate::context::NO_UUID_FLAG;
+use devinfo::{blkid::probe::Probe, DevInfoError};
+
+use std::process::Output;
+use strum_macros::EnumString;
+use tokio::process::Command;
+use tonic::async_trait;
+use tracing::trace;
+
+/// Error type filesystem operations.
+type Error = String;
+
+#[async_trait]
+pub trait FileSystemOps: Send + Sync {
+    /// Create the filesystem using its fs util.
+    async fn create(&self, device: &str) -> Result<(), Error>;
+    /// Set the filesystem uuid using its fs util.
+    async fn set_uuid(&self, device: &str, uuid: &str) -> Result<(), Error>;
+    /// Get the default mount options along with the user passed options for specific filesystems.
+    fn default_mount_flags(&self, mount_flags: Vec<String>) -> Vec<String>;
+}
+
+/// A type to enumerate used filesystems.
+#[derive(EnumString, Clone, Debug, Eq, PartialEq)]
+#[strum(serialize_all = "lowercase")]
+pub enum FileSystem {
+    Ext4,
+    Xfs,
+    DevTmpFs,
+    Unsupported(String),
+}
+
+/// Ext4 filesystem type.
+struct Ext4Fs;
+/// XFS filesystem type.
+struct XFs;
+
+impl FileSystem {
+    /// Convert the top level enumeration to specific filesystem types to access the supported
+    /// operations by them.
+    pub fn to_fs_ops(&self) -> Result<&dyn FileSystemOps, Error> {
+        static EXT4FS: Ext4Fs = Ext4Fs {};
+        static XFS: XFs = XFs {};
+        match self {
+            FileSystem::Ext4 => Ok(&EXT4FS),
+            FileSystem::Xfs => Ok(&XFS),
+            _ => Err(format!("Unsupported filesystem {self}")),
+        }
+    }
+    /// Get a specific filesystem property by specifying the device path.
+    pub fn get_property(device: &str, property: &str) -> Result<String, DevInfoError> {
+        let probe = Probe::new_from_filename(device)?;
+        probe.do_probe()?;
+        probe.lookup_value(property)
+    }
+    /// Get the filesystem specific mount flags that should be used as an default along with used
+    /// defined flags.
+    pub fn default_mount_flags(&self, mount_flags: Vec<String>) -> Result<Vec<String>, Error> {
```

Comment:
Do we need this? Should we let calling code create the ops and then just call default_mount_flags?